plutus-core/src/Language/PlutusCore/Flat.hs

```diff
@@ -82,11 +82,24 @@ implementations for them (if they have any constructors reserved for future use)
 By default, Flat does not use any space to serialise `()`.
 -}
 
+constantWidth :: NumBits
+constantWidth = 3
+
+encodeConstant :: Word8 -> Encoding
+encodeConstant = eBits constantWidth
+
+decodeConstant :: Get Word8
+decodeConstant = dBEBits8 constantWidth
+
 -- See Note [The G, the Tag and the Auto].
 instance Closed uni => Flat (Some (TypeIn uni)) where
-    encode (Some (TypeIn uni)) = encode . map (fromIntegral :: Int -> Word) $ encodeUni uni
+    encode (Some (TypeIn uni)) =
+      encodeListWith encodeConstant .
+        map (fromIntegral :: Int -> Word8) $ encodeUni uni
```

Comment:
`Int` is also platform dependent. Presumably we have an error case here if we need more constant tags than will fit in a `Word8` - does that get handled right?

Reply:
I think the short answer is 'no'.

The longer answer is that the size of the generated term is checked when serialisation occurs, but it is checked up to word boundaries. So there is a failure case where the serialised term uses a couple more bits than it should, but if there are less than 8 bits, the error may not be detected.

I think a fairly good approach would be to test that we don't do bad things at runtime, and if performance suffers a hit we could conditionally compile it on the environment (strip it out of production, keep it in development for running tests).

Shall I add the checks?

Reply:
Well, what I mean is that, for example, you call `fromIntegral` here targeting `Word8`, and I think that will just throw an exception if the `Int` is out of bounds. We could do a check and throw a better error at least.

Reply:
That would be nice, but in the end that `Int` gets turned into a 3 bit value. I think it is far more valuable to check that that `Int` can fit into the 3 bit value.

Reply:
I decided not to use `assert` in the end, as it is disabled by default when using `-O` (so we have to be mindful of one more thing during CI) and it does not provide a good way of printing error messages, and, in this use-case (where we only need to call it once) the location where it is called is obvious.

The only downside is that it is not compiled out of the production library, so it will incur some overhead. If that is a problem we can `#define` it out at some later point (I am running now the benchmarks to see if it impacts us in any significant way, but I suspect it does not).